exporter/elasticsearchexporter/exporter.go

```diff
@@ -298,30 +314,31 @@ func (e *elasticsearchExporter) pushMetricsData(
 					}
 				}
 			}
+		}
+	}
 
-			for fIndex, groupedDataPoints := range groupedDataPointsByIndex {
-				for _, dataPoints := range groupedDataPoints {
-					buf := e.bufferPool.NewPooledBuffer()
-					dynamicTemplates, err := e.model.encodeMetrics(resource, resourceMetric.SchemaUrl(), scope, scopeMetrics.SchemaUrl(), dataPoints, &validationErrs, fIndex, buf.Buffer)
-					if err != nil {
-						buf.Recycle()
-						errs = append(errs, err)
-						continue
-					}
-					if err := session.Add(ctx, fIndex.Index, "", buf, dynamicTemplates); err != nil {
-						// not recycling after Add returns an error as we don't know if it's already recycled
-						if cerr := ctx.Err(); cerr != nil {
-							return cerr
-						}
-						errs = append(errs, err)
-					}
-				}
+	for fIndex, groupedDataPoints := range groupedDataPointsByIndex {
+		for _, dpGroup := range groupedDataPoints {
+			buf := e.bufferPool.NewPooledBuffer()
+			dynamicTemplates, err := e.model.encodeMetrics(
+				dpGroup.resource, dpGroup.resourceSchemaURL, dpGroup.scope, dpGroup.scopeSchemaURL, dpGroup.dataPoints, &validationErrs, fIndex, buf.Buffer)
+			if err != nil {
+				buf.Recycle()
```

Comment:
non-blocker: Calling Recycle like this feels dangerous to me. I feel more comfortable if there's no Recycle method and we don't call recycle on error case, unless we feel the need to also optimize for error path.

Reply:
This isn't new code, I just moved this out of the loop. Can we track this separately?

I don't see why it's dangerous here. We're in control of the code and nothing in the downstream callers holds a reference to the buffer after it returns.